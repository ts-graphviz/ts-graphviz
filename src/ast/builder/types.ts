import type {
  ASTChildNode,
  AttributeASTNode,
  AttributeASTPropaties,
  AttributeListASTNode,
  AttributeListASTPropaties,
  CommentASTNode,
  CommentASTPropaties,
  DotASTNode,
  DotASTPropaties,
  EdgeASTNode,
  EdgeASTPropaties,
  FileRange,
  GraphASTNode,
  GraphASTPropaties,
  LiteralASTNode,
  LiteralASTPropaties,
  NodeASTNode,
  NodeASTPropaties,
  NodeRefASTNode,
  NodeRefASTPropaties,
  NodeRefGroupASTNode,
  NodeRefGroupASTPropaties,
  SubgraphASTNode,
  SubgraphASTPropaties,
} from '../types.js';

export interface BuilderOptions {
  locationFunction: () => FileRange;
}

export interface CreateElement {
  <T extends string>(
    type: 'Literal',
    props: LiteralASTPropaties<T>,
    children: ASTChildNode<LiteralASTNode>[],
  ): LiteralASTNode<T>;
  (type: 'Literal', props: LiteralASTPropaties, children: ASTChildNode<LiteralASTNode>[]): LiteralASTNode;
  (type: 'Dot', props: DotASTPropaties, children: ASTChildNode<DotASTNode>[]): DotASTNode;
  (type: 'Graph', props: GraphASTPropaties, children: ASTChildNode<GraphASTNode>[]): GraphASTNode;
  (type: 'Attribute', props: AttributeASTPropaties, children: ASTChildNode<AttributeASTNode>[]): AttributeASTNode;
  (type: 'Comment', props: CommentASTPropaties, children: ASTChildNode<CommentASTNode>[]): CommentASTNode;
  (
    type: 'AttributeList',
    props: AttributeListASTPropaties,
    children: ASTChildNode<AttributeListASTNode>[],
  ): AttributeListASTNode;
  (type: 'NodeRef', props: NodeRefASTPropaties, children: ASTChildNode<NodeRefASTNode>[]): NodeRefASTNode;
  (
    type: 'NodeRefGroup',
    props: NodeRefGroupASTPropaties,
    children: ASTChildNode<NodeRefGroupASTNode>[],
  ): NodeRefGroupASTNode;
  (type: 'Edge', props: EdgeASTPropaties, children: ASTChildNode<EdgeASTNode>[]): EdgeASTNode;
  (type: 'Node', props: NodeASTPropaties, children: ASTChildNode<NodeASTNode>[]): NodeASTNode;
  (type: 'Subgraph', props: SubgraphASTPropaties, children: ASTChildNode<SubgraphASTNode>[]): SubgraphASTNode;
}

export interface ASTBuilder {
  createElement: CreateElement;
}
