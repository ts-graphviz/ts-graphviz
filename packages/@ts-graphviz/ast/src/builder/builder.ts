import type { FileRange, ASTNode, ASTChildNode } from '../types.js';
import { ASTBuilder, BuilderOptions } from './types.js';

/**
 * Builder is an ASTBuilder that provides a method to create an ASTNode.
 *
 * @group Create AST
 */
export class Builder implements ASTBuilder {
  /**
   * Get the current file range or null
   * @internal
   */
  private getLocation(): FileRange | null {
    return this.options?.locationFunction?.() ?? null;
  }

  /**
   * Constructor of Builder
   * @param options - Options to initialize Builder
   */
  constructor(private options?: Partial<BuilderOptions>) {}

  /**
   * Create an {@link ASTNode} of the specified type
   *
   * @param type - Type of the {@link ASTNode}
   * @param props - Properties of the {@link ASTNode}
   * @param children - Children of the {@link ASTNode}
   * @returns An {@link ASTNode}
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public createElement<T extends ASTNode>(type: T['type'], props: any, children: ASTChildNode<T>[] = []): T {
    return {
      location: this.getLocation(),
      ...props,
      type,
      children,
    };
  }
}
