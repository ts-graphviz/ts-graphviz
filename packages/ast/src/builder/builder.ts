import type { ASTChildNode, ASTNode, FileRange } from '../types.js';
import type { ASTBuilder, BuilderOptions } from './types.js';

/**
 * Builder is an ASTBuilder that provides a method to create an ASTNode.
 *
 * @group Create AST
 */
export class Builder implements ASTBuilder {
  private nodeCount = 0;
  private maxNodes: number;

  /**
   * Get the current file range or null
   * @internal
   */
  private getLocation(): FileRange | null {
    return this.options?.locationFunction?.() ?? null;
  }

  /**
   * Constructor of Builder
   * @param options - Options to initialize Builder
   */
  constructor(private options?: Partial<BuilderOptions>) {
    this.maxNodes = options?.maxASTNodes ?? 100000;
  }

  /**
   * Create an {@link ASTNode} of the specified type
   *
   * @param type - Type of the {@link ASTNode}
   * @param props - Properties of the {@link ASTNode}
   * @param children - Children of the {@link ASTNode}
   * @returns An {@link ASTNode}
   * @throws Error if the maximum number of AST nodes is exceeded
   */
  public createElement<T extends ASTNode>(
    type: T['type'],
    props: any,
    children: ASTChildNode<T>[] = [],
  ): T {
    if (this.maxNodes > 0) {
      this.nodeCount++;
      if (this.nodeCount > this.maxNodes) {
        throw new Error(
          `AST node count (${this.nodeCount}) exceeds maximum allowed (${this.maxNodes}). ` +
            `Consider increasing 'maxASTNodes' option or simplifying the input.`,
        );
      }
    }
    return {
      location: this.getLocation(),
      ...props,
      type,
      children,
    };
  }

  /**
   * Reset the node count. Used internally by the parser between parse invocations.
   * @internal
   */
  public resetNodeCount(): void {
    this.nodeCount = 0;
  }
}
