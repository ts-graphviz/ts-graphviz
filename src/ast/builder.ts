import type { Compass } from '../type/index.js';
import type { AttributeKey } from '../attribute/index.js';
import type {
  DotASTNode,
  StatementASTNode,
  ASTCommonNode,
  LiteralASTNode,
  ClusterStatementASTNode,
  NodeASTNode,
  AttributeASTNode,
  NodeRefASTNode,
  NodeRefGroupASTNode,
  EdgeTargetASTNode,
  AttributeListASTNode,
  CommentASTNode,
  GraphASTNode,
  EdgeASTNode,
  SubgraphASTNode,
  FileRange,
} from './types.js';

export interface BuilderConfig {
  defaultLitelalQuated: boolean | 'html';
  defaultGraphStrict: boolean;
  defaultCommentKind: 'Block' | 'Slash' | 'Macro';
  locationFunction: () => FileRange;
}

export class Builder {
  get defaultLitelalQuated(): boolean | 'html' {
    return this.config.defaultLitelalQuated ?? true;
  }

  get defaultGraphStrict(): boolean {
    return this.config.defaultGraphStrict ?? true;
  }

  get defaultCommentKind(): 'Block' | 'Slash' | 'Macro' {
    return this.config.defaultCommentKind ?? 'Block';
  }

  location(): FileRange {
    return (
      this.config.locationFunction?.() ?? {
        start: {
          offset: NaN,
          line: NaN,
          column: NaN,
        },
        end: {
          offset: NaN,
          line: NaN,
          column: NaN,
        },
      }
    );
  }

  constructor(private config: Partial<BuilderConfig>) {}

  public dot({ body, location = this.location() }: { body: StatementASTNode[] } & Partial<ASTCommonNode>): DotASTNode {
    return {
      type: 'Dot',
      body,
      location,
    };
  }

  public literal<T extends string = string>({
    value,
    quoted = this.defaultLitelalQuated,
    location = this.location(),
  }: {
    value: T;
    quoted?: boolean | 'html';
  } & Partial<ASTCommonNode>): LiteralASTNode<T> {
    return {
      type: 'Literal',
      value,
      quoted,
      location,
    };
  }

  public graph({
    id,
    directed,
    strict = this.defaultGraphStrict,
    body,
    location = this.location(),
  }: {
    id?: LiteralASTNode;
    directed: boolean;
    strict?: boolean;
    body: ClusterStatementASTNode[];
  } & Partial<ASTCommonNode>): GraphASTNode {
    return {
      type: 'Graph',
      id,
      directed,
      strict,
      body,
      location,
    };
  }

  public subgraph({
    id,
    body,
    location = this.location(),
  }: {
    id?: LiteralASTNode;
    body: ClusterStatementASTNode[];
  } & Partial<ASTCommonNode>): SubgraphASTNode {
    return {
      type: 'Subgraph',
      id,
      body,
      location,
    };
  }

  public node({
    id,
    body = [],
    location = this.location(),
  }: {
    body: AttributeASTNode[];
    id: LiteralASTNode;
  } & Partial<ASTCommonNode>): NodeASTNode {
    return {
      type: 'Node',
      id,
      body,
      location,
    };
  }

  public nodeRef({
    id,
    location = this.location(),
    port,
    compass,
  }: {
    id: LiteralASTNode;
    port?: LiteralASTNode;
    compass?: LiteralASTNode<Compass>;
  } & Partial<ASTCommonNode>): NodeRefASTNode {
    return {
      type: 'NodeRef',
      id,
      port,
      compass,
      location,
    };
  }

  public nodeRefGroup({
    body,
    location = this.location(),
  }: { body: NodeRefASTNode[] } & Partial<ASTCommonNode>): NodeRefGroupASTNode {
    return {
      type: 'NodeRefGroup',
      body,
      location,
    };
  }

  public edge({
    targets,
    location = this.location(),
    body,
  }: {
    targets: [from: EdgeTargetASTNode, to: EdgeTargetASTNode, ...rest: EdgeTargetASTNode[]];
    body: AttributeASTNode[];
  } & Partial<ASTCommonNode>): EdgeASTNode {
    return {
      type: 'Edge',
      targets,
      location,
      body,
    };
  }

  public attribute({
    key,
    value,
    location = this.location(),
  }: {
    key: LiteralASTNode<AttributeKey>;
    value: LiteralASTNode;
  } & Partial<ASTCommonNode>): AttributeASTNode {
    return {
      type: 'Attribute',
      key,
      value,
      location,
    };
  }

  public attributes({
    kind,
    body,
    location = this.location(),
  }: {
    kind: 'Graph' | 'Edge' | 'Node';
    body: (AttributeASTNode | CommentASTNode)[];
  } & Partial<ASTCommonNode>): AttributeListASTNode {
    return {
      type: 'AttributeList',
      location,
      kind,
      body,
    };
  }

  public comment({
    kind = this.defaultCommentKind,
    value,
    location = this.location(),
  }: {
    kind?: 'Block' | 'Slash' | 'Macro';
    value: string;
  } & Partial<ASTCommonNode>): CommentASTNode {
    return {
      type: 'Comment',
      kind,
      value,
      location,
    };
  }
}
