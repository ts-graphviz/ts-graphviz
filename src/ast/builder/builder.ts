import type { FileRange, ASTNode, ASTChildNode } from '../types.js';
import { ASTBuilder, BuilderOptions } from './types.js';

export class Builder implements ASTBuilder {
  private getLocation(): FileRange | null {
    return this.options?.locationFunction?.() ?? null;
  }

  constructor(private options?: Partial<BuilderOptions>) {}

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public createElement<T extends ASTNode>(type: T['type'], props: any, children: ASTChildNode<T>[]): T {
    return {
      ...props,
      type,
      location: this.getLocation(),
      children,
    };
  }
}
