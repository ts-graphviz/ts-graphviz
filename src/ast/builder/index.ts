export * from './builder.js';
export * from './types.js';
