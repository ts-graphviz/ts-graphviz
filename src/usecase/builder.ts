import { Graph, GraphAttributes } from '../model/index.js';

interface CreateRootFunction {
  (id?: string, attributes?: GraphAttributes, callback?: (g: Graph) => void): Graph;
  (attributes?: GraphAttributes, callback?: (g: Graph) => void): Graph;
  (id?: string, callback?: (g: Graph) => void): Graph;
  (callback?: (g: Graph) => void): Graph;
}

/** @hidden */
function builder(directed: boolean, strictMode = false): CreateRootFunction {
  function createRoot(id?: string, attributes?: GraphAttributes, callback?: (g: Graph) => void): Graph;
  function createRoot(attributes?: GraphAttributes, callback?: (g: Graph) => void): Graph;
  function createRoot(id?: string, callback?: (g: Graph) => void): Graph;
  function createRoot(callback?: (g: Graph) => void): Graph;
  function createRoot(...args: unknown[]): Graph {
    const id = args.find((arg): arg is string => typeof arg === 'string');
    const attributes = args.find((arg): arg is GraphAttributes => typeof arg === 'object');
    const callback = args.find((arg): arg is (g: Graph) => void => typeof arg === 'function');
    const g = new Graph(directed, id, strictMode, attributes);
    if (typeof callback === 'function') {
      callback(g);
    }
    return g;
  }
  return createRoot;
}

/** API for creating directional graph objects. */
export const digraph = builder(true);

/** API for creating omnidirectional graph objects. */
export const graph = builder(false);

/** Provides a strict mode API. */
export const strict = Object.freeze({
  /** API for creating directional graph objects in strict mode. */
  digraph: builder(true, true),
  /** API for creating omnidirectional graph objects in strict mode. */
  graph: builder(false, true),
});
