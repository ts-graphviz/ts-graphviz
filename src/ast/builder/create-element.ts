import { CreateElement } from './types.js';
import { Builder } from './builder.js';
/**
 * Create an {@link ASTNode} of the specified type
 *
 * @param type - Type of the {@link ASTNode}
 * @param props - Properties of the {@link ASTNode}
 * @param children - Children of the {@link ASTNode}
 * @group Create AST
 * @returns An {@link ASTNode}
 */
export const createElement: CreateElement = Builder.prototype.createElement.bind(new Builder());
