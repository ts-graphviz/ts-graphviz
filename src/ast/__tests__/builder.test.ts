import { AttributeKey } from '../../attribute/index.js';
import { Builder } from '../builder.js';

describe('Builder', () => {
  const builer = new Builder();

  describe('createElement', () => {
    test('Dot', () => {
      const ast = builer.createElement('Dot', {}, []);
      expect(ast).toMatchSnapshot();
    });

    test('Graph', () => {
      const ast = builer.createElement(
        'Graph',
        {
          strict: false,
          directed: true,
        },
        [],
      );
      expect(ast).toMatchSnapshot();
    });

    test('Node', () => {
      const ast = builer.createElement(
        'Node',
        {
          id: builer.createElement(
            'Literal',
            {
              value: 'node_id',
              quoted: false,
            },
            [],
          ),
        },
        [],
      );
      expect(ast).toMatchSnapshot();
    });

    test('Edge', () => {
      const ast = builer.createElement(
        'Edge',
        {
          targets: [
            builer.createElement(
              'NodeRef',
              {
                id: builer.createElement(
                  'Literal',
                  {
                    value: 'a',
                    quoted: true,
                  },
                  [],
                ),
              },
              [],
            ),
            builer.createElement(
              'NodeRef',
              {
                id: builer.createElement(
                  'Literal',
                  {
                    value: 'b',
                    quoted: true,
                  },
                  [],
                ),
              },
              [],
            ),
          ],
        },
        [],
      );
      expect(ast).toMatchSnapshot();
    });

    test('NodeRef', () => {
      const ast = builer.createElement(
        'NodeRef',
        {
          id: builer.createElement(
            'Literal',
            {
              value: 'b',
              quoted: true,
            },
            [],
          ),
        },
        [],
      );
      expect(ast).toMatchSnapshot();
    });

    test('NodeRefGroup', () => {
      const ast = builer.createElement('NodeRefGroup', {}, [
        builer.createElement(
          'NodeRef',
          {
            id: builer.createElement(
              'Literal',
              {
                value: 'a',
                quoted: true,
              },
              [],
            ),
          },
          [],
        ),
        builer.createElement(
          'NodeRef',
          {
            id: builer.createElement(
              'Literal',
              {
                value: 'b',
                quoted: true,
              },
              [],
            ),
          },
          [],
        ),
      ]);
      expect(ast).toMatchSnapshot();
    });

    test('Subgraph', () => {
      const ast = builer.createElement('Subgraph', {}, []);
      expect(ast).toMatchSnapshot();
    });

    test('Literal', () => {
      const ast = builer.createElement(
        'Literal',
        {
          value: 'literal value',
          quoted: true,
        },
        [],
      );
      expect(ast).toMatchSnapshot();
    });

    test('Coment', () => {
      const ast = builer.createElement(
        'Comment',
        {
          kind: 'Macro',
          value: 'This is comment',
        },
        [],
      );
      expect(ast).toMatchSnapshot();
    });

    test('Attribute', () => {
      const ast = builer.createElement(
        'Attribute',
        {
          key: builer.createElement<AttributeKey>(
            'Literal',
            {
              value: 'color',
              quoted: false,
            },
            [],
          ),
          value: builer.createElement(
            'Literal',
            {
              value: 'red',
              quoted: false,
            },
            [],
          ),
        },
        [],
      );
      expect(ast).toMatchSnapshot();
    });
  });
});
