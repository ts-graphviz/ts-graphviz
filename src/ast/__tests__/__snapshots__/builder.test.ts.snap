// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Builder createElement Attribute 1`] = `
Object {
  "children": Array [],
  "key": Object {
    "children": Array [],
    "location": null,
    "quoted": false,
    "type": "Literal",
    "value": "color",
  },
  "location": null,
  "type": "Attribute",
  "value": Object {
    "children": Array [],
    "location": null,
    "quoted": false,
    "type": "Literal",
    "value": "red",
  },
}
`;

exports[`Builder createElement Coment 1`] = `
Object {
  "children": Array [],
  "kind": "Macro",
  "location": null,
  "type": "Comment",
  "value": "This is comment",
}
`;

exports[`Builder createElement Dot 1`] = `
Object {
  "children": Array [],
  "location": null,
  "type": "Dot",
}
`;

exports[`Builder createElement Edge 1`] = `
Object {
  "children": Array [],
  "location": null,
  "targets": Array [
    Object {
      "children": Array [],
      "id": Object {
        "children": Array [],
        "location": null,
        "quoted": true,
        "type": "Literal",
        "value": "a",
      },
      "location": null,
      "type": "NodeRef",
    },
    Object {
      "children": Array [],
      "id": Object {
        "children": Array [],
        "location": null,
        "quoted": true,
        "type": "Literal",
        "value": "b",
      },
      "location": null,
      "type": "NodeRef",
    },
  ],
  "type": "Edge",
}
`;

exports[`Builder createElement Graph 1`] = `
Object {
  "children": Array [],
  "directed": true,
  "location": null,
  "strict": false,
  "type": "Graph",
}
`;

exports[`Builder createElement Literal 1`] = `
Object {
  "children": Array [],
  "location": null,
  "quoted": true,
  "type": "Literal",
  "value": "literal value",
}
`;

exports[`Builder createElement Node 1`] = `
Object {
  "children": Array [],
  "id": Object {
    "children": Array [],
    "location": null,
    "quoted": false,
    "type": "Literal",
    "value": "node_id",
  },
  "location": null,
  "type": "Node",
}
`;

exports[`Builder createElement NodeRef 1`] = `
Object {
  "children": Array [],
  "id": Object {
    "children": Array [],
    "location": null,
    "quoted": true,
    "type": "Literal",
    "value": "b",
  },
  "location": null,
  "type": "NodeRef",
}
`;

exports[`Builder createElement NodeRefGroup 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "id": Object {
        "children": Array [],
        "location": null,
        "quoted": true,
        "type": "Literal",
        "value": "a",
      },
      "location": null,
      "type": "NodeRef",
    },
    Object {
      "children": Array [],
      "id": Object {
        "children": Array [],
        "location": null,
        "quoted": true,
        "type": "Literal",
        "value": "b",
      },
      "location": null,
      "type": "NodeRef",
    },
  ],
  "location": null,
  "type": "NodeRefGroup",
}
`;

exports[`Builder createElement Subgraph 1`] = `
Object {
  "children": Array [],
  "location": null,
  "type": "Subgraph",
}
`;
